lib/puppet/provider/puppet_agent_end_run/puppet_agent_end_run.rb

```diff
@@ -32,8 +32,13 @@ def needs_upgrade?
     return false if desired_version == 'present'
 
     if desired_version == 'latest'
-      latest_version = @resource.catalog.resource('package', 'puppet-agent').parameters[:ensure].latest
-      desired_version = latest_version.match(%r{^(?:[0-9]:)?(\d+\.\d+(\.\d+)?(?:\.\d+))?}).captures.first
+      # Package name might be different to puppet-agent, hence we need to look it up.
+      package_name = @resource.catalog.resource('class', 'puppet_agent')[:package_name]
+
+      # Latest version might be undefined, e.G. if we're about to install a different named
+      # package  than the currently running one. In that case, we'll leave desired_version empty.
+      latest_version = @resource.catalog.resource('package', package_name).parameters[:ensure].latest
+      desired_version = latest_version.match(%r{^(?:[0-9]:)?(\d+\.\d+(\.\d+)?(?:\.\d+))?}).captures.first unless latest_version.nil?
```

Comment:
If `latest_version` is nil, then `desired_version` will be too, and the `versioncmp` function will raise:

```
❯ bundle exec ruby -rpuppet -e 'Puppet::Util::Package.versioncmp(nil, "1.2.3")' 
/home/josh/work/puppet-private/lib/puppet/util/package.rb:12:in `versioncmp': undefined method `scan' for nil:NilClass (NoMethodError)
```

Maybe use empty string instead?

```
desired_version = if latest_version.nil?
                               ""
                             else
                               latest_version.match(%r{^(?:[0-9]:)?(\d+\.\d+(\.\d+)?(?:\.\d+))?}).captures.first
                             end
```

Reply:
If latest_version is nil then desired_version will still be 'latest' - But I think it makes sense to be more explicit here.